java-tools/OsmAndServer/src/main/java/net/osmand/server/controllers/pub/RoutingController.java

```diff
@@ -271,62 +296,72 @@ public ResponseEntity<String> routing(@RequestParam String[] points,
+				List<RouteSegmentResult> res =
+						osmAndMapsService.routing(hhOnlyForce, routeMode, props, list.get(0),
+								list.get(list.size() - 1), list.subList(1, list.size() - 1),
+								avoidRoads == null ? Collections.emptyList() : Arrays.asList(avoidRoads));
 				if (res != null) {
-					routingService.convertResults(resList, features, res);
+					routingService.convertResultsWithElevation(resListElevation, features, res);
 				}
 			} catch (IOException | InterruptedException | RuntimeException e) {
 				LOGGER.error(e.getMessage(), e);
 			}
 		}
-		if (resList.isEmpty()) {
-			resList = new ArrayList<>(list);
-			routingService.calculateStraightLine(resList);
+
+		List<LatLon> resListFallbackLine = new ArrayList<>();
+		boolean reportLimitError = false;
+		if (resListElevation.isEmpty()) {
+			reportLimitError = true;
+			resListFallbackLine = new ArrayList<>(list);
+			routingService.calculateStraightLine(resListFallbackLine);
 			float dist = 0;
-			for (int i = 1; i < resList.size(); i++) {
-				dist += MapUtils.getDistance(resList.get(i - 1), resList.get(i));
+			for (int i = 1; i < resListFallbackLine.size(); i++) {
+				dist += MapUtils.getDistance(resListFallbackLine.get(i - 1), resListFallbackLine.get(i));
 			}
 			props.put("distance", dist);
 		}
-		Feature route = new Feature(Geometry.lineString(resList));
+
+		Feature route = resListElevation.isEmpty() ?
+				new Feature(Geometry.lineString(resListFallbackLine)) :
+				new Feature(Geometry.lineStringElevation(resListElevation));
 		route.properties = props;
 		features.add(0, route);
 		
-		if (tooLong) {
+		if (reportLimitError) {
 			return ResponseEntity.ok(gson.toJson(Map.of("features", new FeatureCollection(features.toArray(new Feature[features.size()])), "msg",
-					MSG_LONG_DIST + maxDist + " km.")));
+					MSG_LONG_DIST + hhOnlyLimit + " km.")));
 		} else {
 			return ResponseEntity.ok(gson.toJson(new FeatureCollection(features.toArray(new Feature[features.size()]))));
 		}
 	}
 	
 	@PostMapping(path = {"/update-route-between-points"}, produces = "application/json")
 	@ResponseBody
-	public ResponseEntity<String> updateRouteBetweenPoints(@RequestParam String start,
-	                                                       @RequestParam String end,
-	                                                       @RequestParam String routeMode,
-	                                                       @RequestParam boolean hasRouting,
-	                                                       @RequestParam int maxDist) throws IOException, InterruptedException {
+	public ResponseEntity<String> updateRouteBetweenPoints(@RequestParam String start, @RequestParam String end,
+	                                                       @RequestParam String routeMode, @RequestParam boolean hasRouting,
+	                                                       @RequestParam(defaultValue = "100") int hhOnlyLimit)
```

Comment:
hh only limit should be server side constant

Reply:
defaultValue here is only to protect against crash if no hhOnlyLimit specified. actually this defaultValue should be never applied because hhOnlyLimit is an always-specified parameter by web-client

Reply:
hhOnlyLimit - shouldn't be passed as parameter

Reply:
2094aa805ed8a0378a8eb42bdce82983861f36a0